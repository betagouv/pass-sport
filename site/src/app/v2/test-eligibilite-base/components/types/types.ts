export enum AGE_RANGE {
  BETWEEN_6_19 = 'BETWEEN_6_19',
  BETWEEN_19_30 = 'BETWEEN_19_30',
  GREATER_THAN_30 = 'GREATER_THAN_30',
}

export enum CHILD_AGE {
  LESS_THAN_SIX = 'LESS_THAN_SIX',
  BTW_SIX_AND_THIRTY = 'BTW_SIX_AND_THIRTY',
  MORE_THAN_THIRTY = 'MORE_THAN_THIRTY',
}
