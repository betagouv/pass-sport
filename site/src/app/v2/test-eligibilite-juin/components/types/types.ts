export enum ALLOWANCE {
  NONE = 'NONE',
  ARS_AEEH_AAH = 'ARS_AEEH_AAH',
  CROUS = 'CROUS',
}
