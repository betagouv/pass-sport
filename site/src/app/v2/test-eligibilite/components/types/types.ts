export enum AGE_RANGE {
  BETWEEN_6_19 = 'BETWEEN_6_19',
  BETWEEN_19_30 = 'BETWEEN_19_30',
  GREATER_THAN_30 = 'GREATER_THAN_30',
}

export enum CHILD_AGE {
  LESS_THAN_SIX = 'lessThanSix',
  BTW_SIX_AND_THIRTY = 'betweenSixAndThrity',
  MORE_THAN_THIRTY = 'moreThanThirty',
}
