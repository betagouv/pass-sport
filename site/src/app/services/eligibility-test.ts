import {
  ConfirmPayload,
  ConfirmResponseBody,
  ConfirmResponseErrorBody,
  SearchPayload,
  SearchResponseBody,
  SearchResponseErrorBody,
} from 'types/EligibilityTest';

import * as Sentry from '@sentry/nextjs';
import { addQrCodeToConfirmResponse } from './qr-code';

export const buildLCAConfirmUrl = (data: ConfirmPayload, isUsingApiV1: boolean): URL => {
  const domain = process.env.LCA_API_URL;

  if (!domain) {
    throw new Error('Error: LCA_API_URL is not set');
  }

  const params = new URLSearchParams();
  params.append('id', data.id);
  params.append('situation', data.situation);
  params.append('organisme', data.organisme);

  const allocataireName = data.recipientLastname;
  if (allocataireName) {
    params.append('allocataireName', allocataireName);
  }

  const allocataireSurname = data.recipientFirstname;
  if (allocataireSurname) {
    params.append('allocataireSurname', allocataireSurname);
  }

  const matricule = data.recipientCafNumber;
  if (matricule) {
    params.append('matricule', matricule);
  }

  const recipientBirthPlace = data.recipientBirthPlace;
  if (recipientBirthPlace) {
    params.append('codeInseeBirth', recipientBirthPlace);
  }

  const recipientBirthDate = data.recipientBirthDate;
  if (recipientBirthDate) {
    params.append('allocataireBirthDate', recipientBirthDate);
  }

  const recipientBirthCountry = data.recipientBirthCountry;
  if (recipientBirthCountry) {
    params.append('codeIso', recipientBirthCountry);
  }

  const baseUrl = isUsingApiV1
    ? `${domain}/gw/psp-server/beneficiaires/confirm`
    : `${domain}/apim/api-asso-admin/passsport/beneficiaires/confirm`;

  const url = new URL(baseUrl);
  url.search = params.toString();

  return url;
};

export const buildLCASearchUrl = (data: SearchPayload, isUsingApiV1: boolean): URL => {
  const domain = process.env.LCA_API_URL;

  if (!domain) {
    throw new Error('Error: LCA_API_URL is not set');
  }

  const params = new URLSearchParams();
  params.append('nom', data.beneficiaryLastname);
  params.append('prenom', data.beneficiaryFirstname);
  params.append('dateNaissance', data.beneficiaryBirthDate);
  params.append('codeInsee', data.recipientResidencePlace);

  const baseUrl = isUsingApiV1
    ? `${domain}/gw/psp-server/beneficiaires/search`
    : `${domain}/apim/api-asso-admin/passsport/beneficiaires/search`;

  const url = new URL(baseUrl);
  url.search = params.toString();

  return url;
};

export const fetchQrCode = async (
  payload: ConfirmPayload,
): Promise<ConfirmResponseBody | ConfirmResponseErrorBody> => {
  const authenticationKey = process.env.LCA_API_KEY;

  if (!authenticationKey) {
    throw new Error('LCA authentication key is missing');
  }

  const url: URL = buildLCAConfirmUrl(payload, true);
  const url2 = buildLCAConfirmUrl(payload, false);

  const [response1, response2] = await Promise.all([
    fetch(url),
    fetch(url2, { headers: { 'X-Gravitee-Api-Key': authenticationKey } }),
  ]);

  if (!response2.ok && !response1.ok) {
    throw new Error(
      `Request to LCA api on /confirm has failed. API V1: Response status is ${response1.status}; Response body is ${JSON.stringify(await response1.json())}. API V2: Response status is ${response2.status}; Response body is ${JSON.stringify(await response2.json())}`,
    );
  }

  const response = !response2.ok ? response1 : response2;
  const responseBody = (await response.json()) as ConfirmResponseBody | ConfirmResponseErrorBody;

  if ('message' in responseBody) {
    Sentry.withScope((scope) => {
      scope.setLevel('warning');
      scope.setExtra('responseBody', responseBody);
      scope.captureMessage('Unexpected response on LCA POST api/eligibility-test/confirm');
    });
    return responseBody;
  }
  if (responseBody instanceof Array && responseBody.length === 0) {
    return responseBody;
  }
  const enhancedResponse = addQrCodeToConfirmResponse(responseBody);
  return enhancedResponse;
};

export const fetchEligible = async (payload: SearchPayload) => {
  const authenticationKey = process.env.LCA_API_KEY;

  if (!authenticationKey) {
    throw new Error('LCA authentication key is missing');
  }

  const url: URL = buildLCASearchUrl(payload, true);
  const url2: URL = buildLCASearchUrl(payload, false);

  const [response1, response2] = await Promise.all([
    fetch(url),
    fetch(url2, { headers: { 'X-Gravitee-Api-Key': authenticationKey } }),
  ]);

  if (!response2.ok && !response1.ok) {
    throw new Error(
      `Request to LCA api on /search has failed. API V1: Response status is ${response1.status}; Response body is ${JSON.stringify(await response1.json())}. API V2: Response status is ${response2.status}; Response body is ${JSON.stringify(await response2.json())}`,
    );
  }

  const response = !response2.ok ? response1 : response2;
  const responseBody = (await response.json()) as SearchResponseBody | SearchResponseErrorBody;

  if ('message' in responseBody) {
    Sentry.withScope((scope) => {
      scope.setLevel('warning');
      scope.setExtra('responseBody', responseBody);
      scope.captureMessage('Unexpected response on LCA POST api/eligibility-test/search');
    });
  }

  return responseBody;
};
