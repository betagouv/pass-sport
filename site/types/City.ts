export interface City {
  nom: string;
  code: string;
  codesPostaux: string[];
}

export interface CityWithPostalCode {
  nom: string;
  code: string;
  postalCode: string;
}
