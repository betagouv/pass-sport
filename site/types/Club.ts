export interface SportGouvJSONResponse {
  results: Club[];
  total_count: number;
}

type YesNo = 'Oui' | 'Non';

interface Club {
  nom: string;
  cp: string;
  activites: string[];
  adresse: string | null;
  com_arm_name: string;
  handicap: YesNo;
  cedex: string | null;
  commune: string | null;
  telephone: string | null;
  courriel: string | null;
  a_accueil_handicap_mental: YesNo | null;
  a_accueil_handicap_moteur: YesNo | null;
  geoloc_finale: {
    lon: number;
    lat: number;
  } | null;
}

export interface Activity {
  activites: string;
}

export interface ActivityResponse {
  results: Activity[];
}
