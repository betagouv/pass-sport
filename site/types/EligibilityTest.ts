type InputState = 'default' | 'success' | 'error';

export interface InputsState {
  beneficiaryLastname: InputState;
  beneficiaryFirstname: InputState;
  beneficiaryBirthDate: InputState;
  recipientResidencePlace: InputState;
}

export interface SearchResponseBody {
  id: number;
  nom: string;
  prenom: string;
  date_naissance: string;
  situation: string;
  organisme: string;
  matricule: string;
}

export interface SearchResponseError {
  message: string;
}
