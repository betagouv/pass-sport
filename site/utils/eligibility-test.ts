import validator from 'validator';

export const isUsingJuneEligibilityTest = new Date().valueOf() > 1717027200000; //2024-05-30T00:00:00.000Z

export const sanitize = (rawInputValue: string): string => {
  return validator.blacklist(rawInputValue.trim(), '\'%;="!?<>:&~#{}()|`^\\[\\]$*ยง,.');
};
