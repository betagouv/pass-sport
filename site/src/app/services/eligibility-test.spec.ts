import {
  buildConfirmPayload,
  buildConfirmResponseBody,
} from '../../../tests/helpers/builders/confirm-response-body';
import { buildLCAConfirmUrl, fetchQrCode } from './eligibility-test';
import { ConfirmResponseBody, ConfirmResponseErrorBody } from '../../../types/EligibilityTest';

global.fetch = jest.fn() as jest.Mock;

jest.mock('./qr-code');

function mockFetch(status: number, responseBody: ConfirmResponseBody | ConfirmResponseErrorBody) {
  (global.fetch as jest.Mock).mockImplementationOnce(() =>
    Promise.resolve({
      json: () => Promise.resolve(responseBody),
      status,
      ok: status === 200,
    }),
  );
}

describe('eligibility-test service', () => {
  describe('buildLCAConfirmUrl', () => {
    const tests = [
      {
        who: 'jeune caf',
        payload: { recipientCafNumber: '1234567' },
        expected: '?id=123456789&situation=jeune&organisme=CAF&matricule=1234567',
      },
      {
        who: 'jeune msa francais',
        payload: {
          recipientLastname: 'fake lastname',
          recipientFirstname: 'fake firstname',
          recipientBirthDate: '01/01/2000',
          recipientBirthPlace: '29260',
          organisme: 'MSA',
        },
        expected:
          '?id=123456789&situation=jeune&organisme=MSA&allocataireName=fake+lastname&allocataireSurname=fake+firstname&codeInseeBirth=29260&allocataireBirthDate=01%2F01%2F2000',
      },
      {
        who: 'jeune msa Ã©tranger',
        payload: {
          recipientLastname: 'fake lastname',
          recipientFirstname: 'fake firstname',
          recipientBirthDate: '01/01/2000',
          recipientBirthCountry: 'DZ',
          organisme: 'MSA',
        },
        expected:
          '?id=123456789&situation=jeune&organisme=MSA&allocataireName=fake+lastname&allocataireSurname=fake+firstname&allocataireBirthDate=01%2F01%2F2000&codeIso=DZ',
      },
      {
        who: 'aah caf',
        payload: { recipientCafNumber: '1234567', situation: 'AAH' },
        expected: '?id=123456789&situation=AAH&organisme=CAF&matricule=1234567',
      },
      {
        who: 'aah msa francais',
        payload: {
          situation: 'AAH',
          organisme: 'MSA',
          recipientBirthPlace: '29260',
        },
        expected: '?id=123456789&situation=AAH&organisme=MSA&codeInseeBirth=29260',
      },
      {
        who: 'aah msa etranger',
        payload: {
          situation: 'AAH',
          organisme: 'MSA',
          recipientBirthCountry: 'DZ',
        },
        expected: '?id=123456789&situation=AAH&organisme=MSA&codeIso=DZ',
      },
    ];
    it.each(tests)("builds appropriate query string for '$who'", ({ payload, expected }) => {
      const queryString = buildLCAConfirmUrl(buildConfirmPayload(payload)).toString();

      const baseUrl = 'http://fake-lca-api-url/gw/psp-server/beneficiaires/confirm';
      expect(queryString).toEqual(`${baseUrl}${expected}`);
    });
  });

  describe('fetchQrCode', () => {
    it('should return an error when remote LCA request is not a 200', async () => {
      const payload = buildConfirmPayload({});

      mockFetch(500, { message: 'error' });

      try {
        await fetchQrCode(payload);
      } catch (error) {
        expect((error as Error).message).toEqual(
          'Request to LCA api on /confirm has failed. Response status is 500. Response body is {"message":"error"}',
        );
      }
    });

    it('should return an empty array when eligible does not exist', async () => {
      const payload = buildConfirmPayload({ recipientCafNumber: '1234567' });

      mockFetch(200, []);

      const data = await fetchQrCode(payload);
      expect(data).toEqual([]);
    });

    it('should return eligible data enhanced with a qrcode url when eligible exists', async () => {
      const payload = buildConfirmPayload({ recipientCafNumber: '1234567' });

      mockFetch(200, buildConfirmResponseBody({}));

      const data = await fetchQrCode(payload);
      expect(data).toMatchSnapshot();
    });
  });
});
