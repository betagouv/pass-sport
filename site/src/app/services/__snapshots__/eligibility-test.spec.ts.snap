// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`eligibility-test service fetchCode should return eligible data enhanced with a code when eligible exists 1`] = `
[
  {
    "a_valider": false,
    "adresse": {
      "code_insee": "29098",
      "code_postal": "29810",
      "commune": "LAMPAUL PLOUARZEL",
      "nom_adresse_postale": "MME DUPOND BABETTE",
      "voie": "31 RUE DU TEST",
    },
    "allocataire": {
      "code_insee_commune_naissance": "38000",
      "code_organisme": "220",
      "commune_naissance": "GRENOBLE",
      "courriel": "fake_email@test.fr",
      "date_naissance": "01/08/1979",
      "matricule": "9999999999999",
      "nom": "DUPOND",
      "prenom": "BABETTE",
      "qualite": "Mme",
      "telephone": "0600000000",
    },
    "created_at": "2024-04-24T14:09:47.646+02:00",
    "date_naissance": "2011-01-01T02:00:00.000+01:00",
    "exercice_id": 4,
    "genre": "F",
    "id": 12967588,
    "id_psp": "24-IIII-IIII",
    "nom": "DUPOND",
    "nom_complet": "MANON DUPOND",
    "organisme": "MSA",
    "prenom": "MANON",
    "refuser": false,
    "situation": "jeune",
    "updated_at": "2023-08-31T12:05:34.000+02:00",
    "uuid_doc": null,
  },
]
`;

exports[`eligibility-test service fetchEligible should return eligible data when eligible exists 1`] = `
[
  {
    "date_naissance": "2011-01-01T23:00:00.000Z",
    "hasMatricule": true,
    "id": 123,
    "nom": "DUPOND",
    "organisme": "MSA",
    "prenom": "MANON",
    "situation": "jeune",
  },
]
`;
